packages/typescript/src/internal/ts_project.bzl

```diff
@@ -0,0 +1,273 @@
+    > to intermittently fail with an error that some outputs were not written.
+    > This is why we depend on
+    > `composite` and/or `incremental` attributes to be provided, so we can tell Bazel to
+    > expect a `.tsbuildinfo` output to ensure it is deleted before a subsequent compilation.
+    > At present, we don't do anything useful with the `.tsbuildinfo` output, and this rule
+    > does not actually have incremental behavior. Deleting the file is actually
+    > counter-productive in terms of TypeScript compile performance.
+    > Follow https://github.com/bazelbuild/rules_nodejs/issues/1726
+
+    > Note: When using Project References, TypeScript will expect to verify that the outputs of referenced
+    > projects are up-to-date with respect to their inputs (this is true even without using the `--build` option).
+    > When using a non-sandboxed spawn strategy, `tsc` can read the sources from other `ts_project`
+    > rules in your project, and will expect that the `tsconfig.json` file for those references will
+    > indicate where the outputs were written. However the `outDir` is determined by this Bazel rule so
+    > it cannot be known from reading the `tsconfig.json` file.
+    > This problem is manifested as a TypeScript diagnostic like
+    > `error TS6305: Output file '/path/to/execroot/a.d.ts' has not been built from source file '/path/to/execroot/a.ts'.`
+    > As a workaround, you can give the Windows "fastbuild" output directory as the `outDir` in your tsconfig file.
+    > On other platforms, the value isn't read so it does no harm.
+    > See https://github.com/bazelbuild/rules_nodejs/tree/master/packages/typescript/test/ts_project as an example.
+    > We hope this will be fixed in a future release of TypeScript;
+    > follow https://github.com/microsoft/TypeScript/issues/37378
+
+    Args:
+        name: A name for the target.
+
+            We recommend you use the basename (no `.json` extension) of the tsconfig file that should be compiled.
+
+        srcs: List of labels of TypeScript source files to be provided to the compiler.
+
+            If absent, defaults to `**/*.ts` (all TypeScript files in the package).
+
+        deps: List of labels of other rules that produce TypeScript typings (.d.ts files)
+
+        tsconfig: Label of the tsconfig.json file to use for the compilation.
+
+            By default, we add `.json` to the `name` attribute.
+
+        extends: List of labels of tsconfig file(s) referenced in `extends` section of tsconfig.
+
+            Must include any tsconfig files "chained" by extends clauses.
+
+        tsc: Label of the TypeScript compiler binary to run.
+
+            Override this if your npm_install or yarn_install isn't named "npm"
+            For example, `tsc = "@my_deps//typescript/bin:tsc"`
+            Or you can pass a custom compiler binary instead.
+
+        declaration: if the `declaration` bit is set in the tsconfig.
```

Comment:
nit: links to typescript docs for these 1:1 mappings for more info?

Reply:
added above